### PR TITLE
Add metric template notebook

### LINKED ISSUE #162: Define Metric collection notebook template
We want to make sure that understanding the metrics we are collecting is straight forward. We also want to provide a template notebook to make contributing additional metrics easy.

The template should provide python code for reading data from the json format output by `/notebooks/data-sources/TestGrid/testgrid_EDA.ipynb`, a simple metric aggregation example, and code that writes the metric to disk and to Ceph as a parquet file.

Before writing a template from scratch please review the existing metric notebooks and see if there is one that can be modified to serve as the template. 

**Acceptance Criteria**    

- [ ] Template notebook for collecting and storing testgrid metrics
- [ ] Add Template to operate first website to  welcome new contributions

### PR DESCRIPTION
**Pull request by Shreyanand**
Co-authored-by: Shrey <shanand@redhat.com>
Co-authored-by: Hema Veeradhi <hveeradh@redhat.com>

## Related Issues and Dependencies
Fixes #162 

## This introduces a breaking change

- [ ] Yes
- [X] No

## Description

This PR creates a notebook that defines a template for calculating metrics using number_of_flakes.ipynb as an example. 

Note: We define all the helper functions here so that they can be reused by just importing this notebook template. If we want the template to not include helper functions, we could move them to a utils.py file. 

Currently, the Pipfile is not locking due to ipython version mismatch.

### CODE REVIEW

**Comment by MichaelClifford** [comment]
:+1:  This is good, @Shreyanand and @hemajv 

Can you also add a section to save the data locally in addition to pushing it to ceph?  So that it lands in the appropriate `data/processed/metrics/<metric_name>/` directory of the repo?

**Comment by Shreyanand** [comment]
>  This is good, @Shreyanand and @hemajv
> 
> Can you also add a section to save the data locally in addition to pushing it to ceph? So that it lands in the appropriate `data/processed/metrics/<metric_name>/` directory of the repo?

Will do! Any pointers on how to fix the pipfile locking issue?

**Comment by MichaelClifford** [comment]
@Shreyanand are you seeing the same error as @oindrillac [here](https://github.com/aicoe-aiops/mailing-list-analysis-toolkit/pull/54#issuecomment-801097682)?

**Comment by Shreyanand** [comment]
> @Shreyanand are you seeing the same error as @oindrillac [here](https://github.com/aicoe-aiops/mailing-list-analysis-toolkit/pull/54#issuecomment-801097682)?

Yes!

**Comment by Shreyanand** [comment]
@MichaelClifford I removed the WIP tag since the locking error may take a while to fix. Please let us know if there other comments apart from saving to disk section and I'll add them as well.

**Comment by chauhankaranraj** [comment]
> Yes!

Ah I've seen this error before too. Can you try explicitly mentioning the version for `black` and then `pipenv install`-ing? I.e.
```
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]
flake8 = "*"
coverage = "*"
Sphinx = "*"

[packages]
awscli = "*"
boto3 = "*"
pandas = "*"
pandas_profiling = "<2.10"
numpy = "*"
scikit-learn = "*"
matplotlib = "*"
seaborn = "*"
ipynb = "*"
beautifulsoup4 = "==4.7.1"
papermill = "*"
python-dotenv = "*"
black = "==20.8b1"

[requires]
python_version = "3.6"
```

**Comment by Shreyanand** [comment]
> black = "==20.8b1"

That worked, thanks man :tada: 
@oindrillac you can try this out as well.

**Review by MichaelClifford — APPROVED** [review]
:clap:

**Review by aakankshaduggal — APPROVED** [review]
Looks good :rocket:

**Comment by MichaelClifford** [comment]
@oindrillac @chauhankaranraj @isabelizimm @SankBad 

Please all take a couple minutes to review this PR. Since this is the template that we should be using to normalize all our existing notebooks this sprint, I want to make sure everyone sees it and has a chance to give their input in before finalizing.

**Review by hemajv — APPROVED** [review]
:ship:

**Review by oindrillac — APPROVED** [review]
LGTM :rocket:

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb [hunk 1]
```diff
@@ -0,0 +1,1052 @@
+{
```
Comment (chauhankaranraj) [inline notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:1]: <p>This groupby would group tests from different tabs and grids into the same group (assuming there are tests whose names don't vary across tabs/grids), and then calculate the % flakes. Is that what we're going for here, or would <code>df.groupby(['tab', 'grid', 'test'])</code> be a more accurate measurement?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aicoe-aiops/ocp-ci-analysis/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (MichaelClifford) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:1]: <p>@<a href="https://github.com/chauhankaranraj" target="_blank">chauhankaranraj</a> I  think here we do want to make a claim about how "flaky" a specific test is across all grids. But good point about noting the difference, perhaps there would be some value in highlight the difference in flake severity for a test across all grids (baseline??) vs a specific grid.    </p>

Comment (Shreyanand) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:1]: <p>I think df.groupby(['tab', 'grid', 'test']) would give flake % for a test over time which is useful and will be calculated in the Superset dashboard. The flake severity metric (Issue https://github.com/aicoe-aiops/ocp-ci-analysis/issues/138)  aims to look at a particular test across tabs and grids to see if it's flaky everywhere or not. Let me know if that seems unreasonable, we could redo the logic and update the issue if required. </p>

Comment (chauhankaranraj) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:1]: > I think here we do want to make a claim about how "flaky" a specific test is across all grids

Gotcha, just wanted to make sure there's no sneaky groupby errors :)

> But good point about noting the difference, perhaps there would be some value in highlight the difference in flake severity for a test across all grids (baseline??) vs a specific grid.

Agreed, it seems worth looking into

**Review by chauhankaranraj — APPROVED** [review]
lgtm :rocket:

File: notebooks/data-sources/TestGrid/metrics/metric_template.ipynb (hunk 1 above)
Comment (chauhankaranraj) [reply notebooks/data-sources/TestGrid/metrics/metric_template.ipynb:1]: > I think df.groupby(['tab', 'grid', 'test']) would give flake % for a test over time which is useful and will be calculated in the Superset dashboard. The flake severity metric (Issue https://github.com/[/issues/138](https://github.com/aicoe-aiops/ocp-ci-analysis/issues/138)) aims to look at a particular test across tabs and grids to see if it's flaky everywhere or not. Let me know if that seems unreasonable, we could redo the logic and update the issue if required.

Yeah that makes sense :+1:

**Comment by isabelizimm** [comment]
lgtm 👍

**Review by SankBad — COMMENTED** [review]
LGTM

**Pull request merged**